pkg/controller/controller_utils.go

```diff
@@ -25,83 +26,97 @@ import (
+// Expectations are a way for controllers to tell the controller manager what they expect. eg:
+//	ControllerExpectations: {
+//		controller1: expects  2 adds in 2 minutes
+//		controller2: expects  2 dels in 2 minutes
+//		controller3: expects -1 adds in 2 minutes => controller3's expectations have already been met
 //	}
 //
 // Implementation:
 //	PodExpectation = pair of atomic counters to track pod creation/deletion
-//	RCExpectationsStore = TTLStore + a PodExpectation per rc
+//	ControllerExpectationsStore = TTLStore + a PodExpectation per controller
 //
 // * Once set expectations can only be lowered
-// * An RC isn't synced till its expectations are either fulfilled, or expire
-// * Rcs that don't set expectations will get woken up for every matching pod
+// * A controller isn't synced till its expectations are either fulfilled, or expire
+// * Controllers that don't set expectations will get woken up for every matching pod
 
-// expKeyFunc to parse out the key from a PodExpectation
-var expKeyFunc = func(obj interface{}) (string, error) {
+// ExpKeyFunc to parse out the key from a PodExpectation
+var ExpKeyFunc = func(obj interface{}) (string, error) {
 	if e, ok := obj.(*PodExpectations); ok {
 		return e.key, nil
 	}
 	return "", fmt.Errorf("Could not find key for obj %#v", obj)
 }
 
-// RCExpectationsManager is an interface that allows users to set and wait on expectations.
+// ControllerExpectationsInterface is an interface that allows users to set and wait on expectations.
 // Only abstracted out for testing.
-type RCExpectationsManager interface {
-	GetExpectations(rc *api.ReplicationController) (*PodExpectations, bool, error)
-	SatisfiedExpectations(rc *api.ReplicationController) bool
-	DeleteExpectations(rcKey string)
-	ExpectCreations(rc *api.ReplicationController, adds int) error
-	ExpectDeletions(rc *api.ReplicationController, dels int) error
-	CreationObserved(rc *api.ReplicationController)
-	DeletionObserved(rc *api.ReplicationController)
+type ControllerExpectationsInterface interface {
+	GetExpectations(controllerKey string) (*PodExpectations, bool, error)
+	SatisfiedExpectations(controllerKey string) bool
+	DeleteExpectations(controllerKey string)
+	SetExpectations(controllerKey string, add, del int) error
+	ExpectCreations(controllerKey string, adds int) error
+	ExpectDeletions(controllerKey string, dels int) error
+	CreationObserved(controllerKey string)
+	DeletionObserved(controllerKey string)
 }
 
-// RCExpectations is a ttl cache mapping rcs to what they expect to see before being woken up for a sync.
-type RCExpectations struct {
+// ControllerExpectations is a ttl cache mapping controllers to what they expect to see before being woken up for a sync.
+type ControllerExpectations struct {
 	cache.Store
 }
 
-// GetExpectations returns the PodExpectations of the given rc.
-func (r *RCExpectations) GetExpectations(rc *api.ReplicationController) (*PodExpectations, bool, error) {
-	rcKey, err := rcKeyFunc(rc)
```

Comment:
I moved the call to KeyFunc out of these methods mainly so that Expectations can be used with objects besides Replication Controllers.

Another option is to change api.ReplicationController to runtime.Object (I think?). I thought separating the key function from the Expectations data structure modularized the code better, but a drawback is that we have to do more error checking.

Let me know what you think!